flink-kubernetes/src/main/java/org/apache/flink/kubernetes/KubernetesResourceManager.java

```diff
@@ -0,0 +1,381 @@
+
+		this.kubeClient = createFlinkKubeClient();
+
+		try {
+			getWorkerNodesFromPreviousAttempts();
+		} catch (Exception e) {
+			throw new ResourceManagerException(e);
+		}
+
+		this.kubeClient.watchPodsAndDoCallback(getTaskManagerLabels(), this);
+	}
+
+	@Override
+	public CompletableFuture<Void> onStop() {
+		// shut down all components
+		Throwable firstException = null;
+
+		if (kubeClient != null) {
+			try {
+				kubeClient.close();
+			} catch (Throwable t) {
+				firstException = t;
+			}
+		}
+
+		final CompletableFuture<Void> terminationFuture = super.onStop();
+
+		if (firstException != null) {
+			return FutureUtils.completedExceptionally(new FlinkException(
+				"Error while shutting down Kubernetes resource manager", firstException));
+		} else {
+			return terminationFuture;
+		}
+	}
+
+	@Override
+	protected void internalDeregisterApplication(ApplicationStatus finalStatus, @Nullable String optionalDiagnostics) {
+		LOG.info("Stopping kubernetes cluster, id: {}", clusterId);
+		this.kubeClient.stopAndCleanupCluster(this.clusterId);
+	}
+
+	@Override
+	public Collection<ResourceProfile> startNewWorker(ResourceProfile resourceProfile) {
+		LOG.info("Starting new worker with resource profile, {}", resourceProfile.toString());
+		if (!slotsPerWorker.iterator().next().isMatching(resourceProfile)) {
+			return Collections.emptyList();
+		}
+		requestKubernetesPod();
+		return slotsPerWorker;
+	}
+
+	@Override
+	protected KubernetesWorkerNode workerStarted(ResourceID resourceID) {
+		return workerNodeMap.get(resourceID);
+	}
+
+	@Override
+	public boolean stopWorker(final KubernetesWorkerNode worker) {
+		Preconditions.checkNotNull(this.kubeClient);
+		LOG.info("Stopping Worker {}.", worker.getResourceID().toString());
+		try {
+			this.kubeClient.stopPod(worker.getResourceID().toString());
+		} catch (Exception e) {
+			this.kubeClient.handleException(e);
+			return false;
+		}
+		workerNodeMap.remove(worker.getResourceID());
+		return true;
+	}
+
+	@Override
+	public void onAdded(List<KubernetesPod> pods) {
+		runAsync(() -> {
+			for (KubernetesPod pod : pods) {
+				log.info("Received new TaskExecutor pod: {} - Remaining pending pod requests: {}",
+					pod.getName(), numPendingPodRequests);
+
+				if (numPendingPodRequests > 0) {
+					numPendingPodRequests--;
+					final KubernetesWorkerNode worker = new KubernetesWorkerNode(new ResourceID(pod.getName()));
+					workerNodeMap.putIfAbsent(worker.getResourceID(), worker);
+				}
+			}
+		});
+	}
+
+	@Override
+	public void onModified(List<KubernetesPod> pods) {
+		runAsync(() -> pods.forEach(this::removePodIfTerminated));
+	}
+
+	@Override
+	public void onDeleted(List<KubernetesPod> pods) {
+		runAsync(() -> pods.forEach(this::removePodIfTerminated));
+	}
+
+	@Override
+	public void onError(List<KubernetesPod> pods) {
+		runAsync(() -> pods.forEach(this::removePodIfTerminated));
+	}
+
+	@VisibleForTesting
+	ConcurrentMap<ResourceID, KubernetesWorkerNode> getWorkerNodeMap() {
```

Comment:
getWorkerNodeMap -> getAllWorkerNodes/getWorkerNodes

Reply:
Will fix it.